eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/RouteBlinding.scala

```diff
@@ -66,26 +121,64 @@ object RouteBlindingEncryptedDataCodecs {
   private val outgoingNodeId: Codec[OutgoingNodeId] = (("length" | constant(hex"21")) :: ("node_id" | publicKey)).as[OutgoingNodeId]
   private val pathId: Codec[PathId] = variableSizeBytesLong(varintoverflow, "path_id" | bytes).as[PathId]
   private val nextBlinding: Codec[NextBlinding] = (("length" | constant(hex"21")) :: ("blinding" | publicKey)).as[NextBlinding]
+  private val paymentRelay: Codec[PaymentRelay] = (
+    ("length" | constant(hex"0a")) ::
+    ("fee_base_msat" | millisatoshi32) ::
+      ("fee_proportional_millionths" | uint32) ::
+      ("cltv_expiry_delta" | cltvExpiryDelta)).as[PaymentRelay]
+  private val paymentConstraints: Codec[PaymentConstraints] = variableSizeBytesLong(varintoverflow,
+    ("max_cltv_expiry" | cltvExpiry) ::
+      ("htlc_minimum_msat" | millisatoshi) ::
+      ("allowed_features" | featuresCodec)).as[PaymentConstraints]
+
+
+  private val messageRelayDataTlvCodec = discriminated[RouteBlindingEncryptedDataTlv].by(varint)
+    .typecase(UInt64(1), padding)
+    .typecase(UInt64(4), outgoingNodeId)
+    .typecase(UInt64(8), nextBlinding)
+
+  val messageRelayDataCodec: Codec[MessageRelayData] = TlvCodecs.tlvStream[RouteBlindingEncryptedDataTlv](messageRelayDataTlvCodec).complete.narrow({
+    case tlvs if tlvs.get[OutgoingNodeId].isEmpty => Attempt.failure(MissingRequiredTlv(UInt64(4)))
+    case tlvs => Attempt.successful(MessageRelayData(tlvs))
+  },{
+    case MessageRelayData(tlvs) => tlvs
+  })
 
-  private val encryptedDataTlvCodec = discriminated[RouteBlindingEncryptedDataTlv].by(varint)
+  private val paymentRelayDataTlvCodec = discriminated[RouteBlindingEncryptedDataTlv].by(varint)
     .typecase(UInt64(1), padding)
     .typecase(UInt64(2), outgoingChannelId)
-    .typecase(UInt64(4), outgoingNodeId)
-    .typecase(UInt64(6), pathId)
     .typecase(UInt64(8), nextBlinding)
+    .typecase(UInt64(10), paymentRelay)
+    .typecase(UInt64(12), paymentConstraints)
+
+  val paymentRelayDataCodec: Codec[PaymentRelayData] = TlvCodecs.tlvStream[RouteBlindingEncryptedDataTlv](paymentRelayDataTlvCodec).complete.narrow({
+    case tlvs if tlvs.get[OutgoingChannelId].isEmpty => Attempt.failure(MissingRequiredTlv(UInt64(2)))
+    case tlvs if tlvs.get[PaymentRelay].isEmpty => Attempt.failure(MissingRequiredTlv(UInt64(10)))
+    case tlvs => Attempt.successful(PaymentRelayData(tlvs))
+  }, {
+    case PaymentRelayData(tlvs) => tlvs
+  })
+
+  private val finalRecipientDataTlvCodec = discriminated[RouteBlindingEncryptedDataTlv].by(varint)
+    .typecase(UInt64(1), padding)
+    .typecase(UInt64(6), pathId)
+    .typecase(UInt64(12), paymentConstraints)
```

Comment:
This has the same issue that the `BlindOnionPaymentPayloadTlv` trait in `PaymentOnion.scala` has: you're encoding requirements in the codec (by removing the tlvs that shouldn't be provided for the final payload), but you're not actually encoding them all, because a final payload for a message shouldn't have a `paymentConstraints` field...

Similarly to my comment in `PaymentOnion.scala`, I think we should make the codecs and type hierarchy simpler here and instead do the validation in higher layers. At first I wouldn't even do that validation, if other tlvs are provided we can simply ignore them. Moreover, for the final payload, we generated that payload, so we know it cannot contain any unwanted tlv!

Reply:
> because a final payload for a message shouldn't have a `paymentConstraints` field...

Why not? Actually, having `paymentConstraints` in the intermediary hops is just a way to make the route fail early but we shouldn't trust it, the important `paymentConstraints` is the one in the final payload.

> Moreover, for the final payload, we generated that payload, so we know it cannot contain any unwanted tlv!

No, anyone that has our public key can generate such a payload, it's not necessarily us.

Reply:
> Why not? Actually, having paymentConstraints in the intermediary hops is just a way to make the route fail early but we shouldn't trust it, the important paymentConstraints is the one in the final payload.

You're completely right, I got confused between payment constraints and payment info!

> No, anyone that has our public key can generate such a payload, it's not necessarily us.

Good point as well! We will validate the `pathId` though at a higher layer, so in practice this isn't an issue, is it?